### DISCUSSION TITLE
Exposing `ModelField` to `__modify_schema__()`?

### DISCUSSION THREAD

**Discussion by antonagestam** [pydantic/pydantic#2902] [Feature Request]
Hi!

I am working on implementing support for pydantic in [phantom-types](https://github.com/antonagestam/phantom-types). A library for creating arbitrarily narrow types on top of builtin types, while still using builtin types for runtime representation.

The Custom Types support in pydantic has mostly made this straight-forward but I'm bumping into an issue when trying to implement `__modify_schema__` for sized collection types. These are types that constrain the size of a collection, for instance [the `NonEmpty` type](https://github.com/antonagestam/phantom-types/blob/main/phantom/sized.py#L80) doesn't allow empty collections. 

When combining these with pydantic, a typical usage could like this, creating a model with a field mandating at least one value:

```python
from phantom.sized import NonEmpty
from pydantic import BaseModel

class SomeModel(BaseModel):
    values: NonEmpty[int]
```

The problem is that the type argument (or the ModelField itself) isn't exposed to `__modify_schema__`, so currently a user would have to create custom subtypes of `NonEmpty` manually adding the value type for each usage.

Is there already some mechanism for solving this, or is it something that could be addressed by for instance passing the `ModelField` to `__modify_schema__` so that `get_args()` could be used to extract the type arguments from `ModelField.outer_type_`? (Or `ModelField.sub_fields`?)

For reference, here is an issue on the phantom-types project to track this: https://github.com/antonagestam/phantom-types/issues/111

**Comment by jasujm** [comment]
Thank you for the amazing library!

I have a similar use case, where generating schema for generic type would need access to subfields. A solution like the one described by above would be much appreciated!

Ideally, modifying schemas would be as flexible as writing validators (where you get all sorts of context variables by defining them in signature).

**Comment by jasujm** [comment]
I opened draft PR to address the issue: https://github.com/samuelcolvin/pydantic/pull/3434
